song_rater/views.py

```diff
@@ -40,23 +44,46 @@ def get_context_data(self, **kwargs):
         return context
 
 
-class SongDetailView(DetailView):
+class SongDisplay(DetailView):
     model = Song
 
     def get_context_data(self, **kwargs):
         context = super().get_context_data(**kwargs)
+        context['rating_list'] = context['song'].ratings.all()
+        context['form'] = RatingForm(initial={'post': self.object})
 
-        rating_list = []
+        return context
 
-        for rating in context['song'].ratings.all():
-            rating_list.append({
-                'created': rating.created,
-                'rating': rating.rating,
-            })
 
-        context['rating_list'] = rating_list
+class SongInterest(SingleObjectMixin, FormView):
+    template_name = 'song_rater/song_detail.html'
+    form_class = RatingForm
+    model = Song
+
+    def get_success_url(self):
+        return reverse('song_rater:song-detail', kwargs={'pk': self.object.pk})
 
-        return context
+    def post(self, request, *args, **kwargs):
+        self.object = self.get_object()
+        return super().post(request, *args, **kwargs)
+
+    def form_valid(self, form):
+        model_instance = form.save(commit=False)
+        model_instance.song = self.object
+        model_instance.save()
+
+        return super().form_valid(form)
+
+
+class SongDetail(View):
+
+    def get(self, request, *args, **kwargs):
+        view = SongDisplay.as_view()
+        return view(request, *args, **kwargs)
+
+    def post(self, request, *args, **kwargs):
+        view = SongInterest.as_view()
+        return view(request, *args, **kwargs)
```

Comment:
Als een `SongDisplay` niet altijd al een liedje laat zien met rating en form, zouden we dan niet een klasse kunnen maken die het liedje + ratings + form laat zien op een `GET` en het resultaat verwerkt op een `POST`?

Reply:
Yes, I combined the functionality of the different views to one view. Initially I found it hard to combine these two functionalities in a good way, but with some puzzling it works well now.